enclave/asym_keys.c

```diff
@@ -0,0 +1,553 @@
+        OE_RAISE(result);
+
+    /* Allocate and call again. */
+    key_buffer_local = (uint8_t*)oe_malloc(key_buffer_size_local);
+    if (key_buffer_local == NULL)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    result = oe_get_seal_key(
+        key_info, key_info_size, key_buffer_local, &key_buffer_size_local);
+
+    if (result != OE_OK)
+        OE_RAISE(result);
+
+    result = OE_OK;
+    *key_buffer = key_buffer_local;
+    *key_buffer_size = key_buffer_size_local;
+    key_buffer_local = NULL;
+
+done:
+    if (key_buffer_local != NULL)
+    {
+        oe_secure_zero_fill(key_buffer_local, key_buffer_size_local);
+        oe_free(key_buffer_local);
+    }
+
+    return result;
+}
+
+static oe_result_t _create_asymmetric_keypair(
+    const oe_asymmetric_key_params_t* key_params,
+    const uint8_t* master_key,
+    size_t master_key_size,
+    oe_ec_private_key_t* private_key,
+    oe_ec_public_key_t* public_key)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    uint8_t key[OE_SHA256_SIZE];
+    uint8_t key_prev[sizeof(key)];
+    oe_ec_type_t type;
+
+    if (!key_params || !master_key || !private_key || !public_key)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    switch (key_params->type)
+    {
+        case OE_ASYMMETRIC_KEY_EC_SECP256P1:
+            type = OE_EC_TYPE_SECP256R1;
+            break;
+        default:
+            OE_RAISE(OE_INVALID_PARAMETER);
+    }
+
+    /* First, derive a key from the given key. */
+    OE_CHECK(
+        oe_kdf_derive_key(
+            OE_KDF_HMAC_SHA256_CTR,
+            master_key,
+            master_key_size,
+            key_params->user_data,
+            key_params->user_data_size,
+            key,
+            sizeof(key)));
+
+    /*
+     * If the derived key is not a valid ECC private key, then we use the
+     * derived key to derive another candidate key. The main requirement
+     * for the derived key is to be between [1, N-1], where N is the order
+     * (number of points) on the elliptic curve. For most curves, it's
+     * very likely that derived key is already within that range.
+     */
+    while (!oe_ec_valid_raw_private_key(type, key, sizeof(key)))
+    {
+        /* Copy to a temporary buffer before doing the key derivation. */
+        OE_CHECK(oe_memcpy_s(key_prev, sizeof(key_prev), key, sizeof(key)));
+
+        /* Derive a new key. */
+        OE_CHECK(
+            oe_kdf_derive_key(
+                OE_KDF_HMAC_SHA256_CTR,
+                key_prev,
+                sizeof(key_prev),
+                key_params->user_data,
+                key_params->user_data_size,
+                key,
+                sizeof(key)));
+    }
+
+    /* Derive both public and private keys. */
+    OE_CHECK(
+        oe_ec_generate_key_pair_from_private(
+            type, key, sizeof(key), private_key, public_key));
+
+    result = OE_OK;
+
+done:
+    oe_secure_zero_fill(key, sizeof(key));
+    oe_secure_zero_fill(key_prev, sizeof(key_prev));
+    return result;
```

Comment:
These two buffers are all inside the enclave's stack, do they need to be zeroed out to prevent any leak?

Reply:
Yeah it's not really necessary since the enclave can only see its own memory, but it's just good practice.